### DISCUSSION TITLE
Spinning wheel when "heavy" requests happens

### DISCUSSION THREAD

**Discussion by easyservices** [moneymanagerex/moneymanagerex#6117] [Ideas]
Hi there,

I would like to propose a feature request : my  database is quite large and when I use the advanced filtering capabilities based on amounts the user interface hangs for a while the time for the db query to complete. It is "disturbing" for end-users as they might think that the app is "frozen" (which it is) and will not respond anymore. Therefore I propose a king of "spinning wheel" or any message that can pop up during that time (but without a button to close it as it is annoying) and will close down by itself when the query/user interface is filled up with the results.

**Comment by tactilis** [comment]
> my database is quite large and when I use the advanced filtering capabilities based on amounts the user interface hangs for a while

1. Could you let us know what the `Total Transactions` count is in the Transaction Statistics section of the Dashboard please?

2. Could you take a screenshot (Alt+Prt Sc key) of the Transaction Filter dialog and paste it here so we can see the settings that produce a delay before returning the results.

**Reply by easyservices** [reply]
It is impressive how fast you are in the reply :-) Amazing! Thanks a lot for your support.

**Reply by tactilis** [reply]
> It is impressive how fast you are in the reply :-)

Thanks. And note your query and my reply is going over a network 😉

**Comment by easyservices** [comment]
1. I have 13.286 total transactions
2. Please see below the screenshot.
![Capture d’écran 2023-07-17 163150](https://github.com/moneymanagerex/moneymanagerex/assets/19292451/7fa14cef-0e78-4f80-8651-c7c92a144253)

Ps : I forgot to mention that I'm running the fatclient on my workstation but opens the database from a LAN (smb file share) and now that I think about...it should explain the long running transaction time to execute the query over the network. However it might make sense to implement a "wait screen" just to make sure that the end-user will not "forcely" shutdown the app thinking it is frozen (bugged).

**Comment by tactilis** [comment]
Use of SMB file share noted.

I've just created a test database containing 30,000 transactions with a range of withdrawal and deposit values.
Filtering on an Amount Range of 15.00 to 100.00 generates 9,800 records and takes approx 0.5 seconds on my laptop from the moment of clicking OK to the the transaction list redraw being completed (measured on a screen video).

As a test to understand how much impact the network is having, could you make a copy of the .mmb database locally on your workstation (give the file a different name perhaps) and try the filtering again. How long does it take now?

And how long does the same filtering actually take over the network?

<br>

A wait screen (or a busy-cursor) might be a good idea - but really all it shows is that the application popped up the dialog (or changed the cursor). It doesn't prove that work is actually taking place and, if the wait is too long, then the user may terminate the app anyway.

**Reply by easyservices** [reply]
Very interesting indeed. I did perform the same advanced filtering over the LAN and then by copying the database to my workstation...it takes the same duration for the query to run until the screen is refreshed : about 10 sec.

Very bizarre as my workstation is rock solid, runs on a SSD,1 16Gb RAM and a decent CPU Intel Core I7 9700F 3GHz with a good score on CPU benchmark.

I also perform on a regular basis the "DB optimization" within MoneyEx (1 time every 3 months).

**Reply by easyservices** [reply]
I understand your point on the pop dialog however I always prefer to know that something happens than waiting without being informed and as soon as you move the mouse you can see that the app screen is frozen like in the 90's when you were running the UI and the business logic within a single thread, the one locking the other...

**Reply by easyservices** [reply]
For info I am on W11 and I had the same issue on W10 as well. It is a gaming machine so do not ask me to run Linux please.

**Reply by tactilis** [reply]
> It is a gaming machine so do not ask me to run Linux please

😄 You won't get any OS wars here.

<br>

> Very interesting indeed. I did perform the same advanced filtering over the LAN and then by copying the database to my workstation...it takes the same duration for the query to run until the screen is refreshed : about 10 sec.

That's an outrageously long period of time. No wonder you have raised this issue. The query and refresh should never take that long.

Your PC workstation has a higher specification than my laptop and yet the performance I see, and with many more transactions, is much faster than you are getting.

At least we have ruled out your LAN as being the problem.

Would you be willing to share (via email) a copy of your database? It would be deleted after investigation.

**Reply by easyservices** [reply]
Unfortunatly I can't send you my database as you will get all my bank accounts numbers and plenty of personal notes for some of the transactions! It is too confidential to share even if you delete it afterwards. Anonymizing all of them will take a lot of time...
I did another test but this time on my wife's laptop with both DB on LAN DB and the locally copied. It takes 18s for the same search.

My 10 cents: I use MMEX from a very long time and I came across all of the minor and major upgrades (btw I have the V1.6.4 64 bits) and it sounds reasonable to think that an index is missing in the DB as such a high duration for a query might be driven by scanning each row instead of using an index.

Below my current DB structure and indexing. It might help...

--
-- File generated with SQLiteStudio v3.2.1 on mar. juil. 18 10:57:52 2023
--
-- Text encoding used: System
--
PRAGMA foreign_keys = off;
BEGIN TRANSACTION;

-- Table: ACCOUNTLIST_V1
CREATE TABLE ACCOUNTLIST_V1 (
    ACCOUNTID       INTEGER PRIMARY KEY,
    ACCOUNTNAME     TEXT    NOT NULL,
    ACCOUNTTYPE     TEXT    NOT NULL,
    ACCOUNTNUM      TEXT,
    STATUS          TEXT    NOT NULL,
    NOTES           TEXT,
    HELDAT          TEXT,
    WEBSITE         TEXT,
    CONTACTINFO     TEXT,
    ACCESSINFO      TEXT,
    INITIALBAL      NUMERIC,
    FAVORITEACCT    TEXT    NOT NULL,
    CURRENCYID      NUMERIC NOT NULL,
    STATEMENTLOCKED INTEGER,
    STATEMENTDATE   TEXT,
    MINIMUMBALANCE  NUMERIC,
    CREDITLIMIT     NUMERIC,
    INTERESTRATE    NUMERIC,
    PAYMENTDUEDATE  TEXT,
    MINIMUMPAYMENT  NUMERIC,
    INITIALDATE     TEXT
);


-- Table: ASSETS_V1
CREATE TABLE ASSETS_V1 (
    ASSETID         INTEGER PRIMARY KEY,
    STARTDATE       TEXT    NOT NULL,
    ASSETNAME       TEXT,
    VALUE           NUMERIC,
    VALUECHANGE     TEXT,
    NOTES           TEXT,
    VALUECHANGERATE NUMERIC,
    ASSETTYPE       TEXT,
    ASSETSTATUS     TEXT,
    CURRENCYID      INTEGER,
    VALUECHANGEMODE TEXT
);


-- Table: ATTACHMENT_V1
CREATE TABLE ATTACHMENT_V1 (
    ATTACHMENTID INTEGER NOT NULL
                         PRIMARY KEY,
    REFTYPE      TEXT    NOT NULL,
    REFID        INTEGER NOT NULL,
    DESCRIPTION  TEXT    COLLATE NOCASE,
    FILENAME     TEXT    NOT NULL
                         COLLATE NOCASE
);


-- Table: BILLSDEPOSITS_V1
CREATE TABLE BILLSDEPOSITS_V1 (
    BDID               INTEGER PRIMARY KEY,
    ACCOUNTID          NUMERIC NOT NULL,
    TOACCOUNTID        NUMERIC,
    PAYEEID            NUMERIC NOT NULL,
    TRANSCODE          TEXT    NOT NULL,
    TRANSAMOUNT        NUMERIC NOT NULL,
    STATUS             TEXT,
    TRANSACTIONNUMBER  TEXT,
    NOTES              TEXT,
    CATEGID            NUMERIC,
    TRANSDATE          TEXT,
    FOLLOWUPID         NUMERIC,
    TOTRANSAMOUNT      NUMERIC,
    REPEATS            NUMERIC,
    NEXTOCCURRENCEDATE TEXT,
    NUMOCCURRENCES     NUMERIC
);


-- Table: BUDGETSPLITTRANSACTIONS_V1
CREATE TABLE BUDGETSPLITTRANSACTIONS_V1 (
    SPLITTRANSID     INTEGER PRIMARY KEY,
    TRANSID          NUMERIC NOT NULL,
    CATEGID          NUMERIC,
    SPLITTRANSAMOUNT NUMERIC,
    NOTES            TEXT
);


-- Table: BUDGETTABLE_V1
CREATE TABLE BUDGETTABLE_V1 (
    BUDGETENTRYID INTEGER PRIMARY KEY,
    BUDGETYEARID  NUMERIC,
    CATEGID       NUMERIC,
    PERIOD        TEXT    NOT NULL,
    AMOUNT        NUMERIC NOT NULL,
    NOTES         TEXT,
    ACTIVE        INTEGER
);


-- Table: BUDGETYEAR_V1
CREATE TABLE BUDGETYEAR_V1 (
    BUDGETYEARID   INTEGER PRIMARY KEY,
    BUDGETYEARNAME TEXT    NOT NULL
);


-- Table: CATEGORY_V1
CREATE TABLE CATEGORY_V1 (
    CATEGID   INTEGER PRIMARY KEY,
    CATEGNAME TEXT    NOT NULL
                      COLLATE NOCASE,
    ACTIVE    INTEGER,
    PARENTID  INTEGER,
    UNIQUE (
        CATEGNAME,
        PARENTID
    )
);


-- Table: CHECKINGACCOUNT_V1
CREATE TABLE CHECKINGACCOUNT_V1 (
    TRANSID           INTEGER PRIMARY KEY,
    ACCOUNTID         NUMERIC NOT NULL,
    TOACCOUNTID       NUMERIC,
    PAYEEID           NUMERIC NOT NULL,
    TRANSCODE         TEXT    NOT NULL,
    TRANSAMOUNT       NUMERIC NOT NULL,
    STATUS            TEXT,
    TRANSACTIONNUMBER TEXT,
    NOTES             TEXT,
    CATEGID           NUMERIC,
    TRANSDATE         TEXT,
    FOLLOWUPID        NUMERIC,
    TOTRANSAMOUNT     NUMERIC,
    LASTUPDATEDTIME   TEXT,
    DELETEDTIME       TEXT
);


-- Table: CURRENCYFORMATS_V1
CREATE TABLE CURRENCYFORMATS_V1 (
    CURRENCYID      INTEGER PRIMARY KEY,
    CURRENCYNAME    TEXT    NOT NULL,
    PFX_SYMBOL      TEXT,
    SFX_SYMBOL      TEXT,
    DECIMAL_POINT   TEXT,
    GROUP_SEPARATOR TEXT,
    UNIT_NAME       TEXT,
    CENT_NAME       TEXT,
    SCALE           NUMERIC,
    BASECONVRATE    NUMERIC,
    CURRENCY_SYMBOL TEXT,
    CURRENCY_TYPE   TEXT
);


-- Table: CURRENCYHISTORY_V1
CREATE TABLE CURRENCYHISTORY_V1 (
    CURRHISTID  INTEGER PRIMARY KEY,
    CURRENCYID  INTEGER NOT NULL,
    CURRDATE    TEXT    NOT NULL,
    CURRVALUE   NUMERIC NOT NULL,
    CURRUPDTYPE INTEGER,
    UNIQUE (
        CURRENCYID,
        CURRDATE
    )
);


-- Table: CUSTOMFIELD_V1
CREATE TABLE CUSTOMFIELD_V1 (
    FIELDID     INTEGER NOT NULL
                        PRIMARY KEY,
    REFTYPE     TEXT    NOT NULL/* Transaction, Stock, Asset, BankAccount, RepeatingTransaction, Payee */,
    DESCRIPTION TEXT    COLLATE NOCASE,
    TYPE        TEXT    NOT NULL/* String, Integer, Decimal, Boolean, Date, Time, SingleChoice, MultiChoice */,
    PROPERTIES  TEXT    NOT NULL
);


-- Table: CUSTOMFIELDDATA_V1
CREATE TABLE CUSTOMFIELDDATA_V1 (
    FIELDATADID INTEGER NOT NULL
                        PRIMARY KEY,
    FIELDID     INTEGER NOT NULL,
    REFID       INTEGER NOT NULL,
    CONTENT     TEXT,
    UNIQUE (
        FIELDID,
        REFID
    )
);


-- Table: INFOTABLE_V1
CREATE TABLE INFOTABLE_V1 (
    INFOID    INTEGER NOT NULL
                      PRIMARY KEY,
    INFONAME  TEXT    NOT NULL,
    INFOVALUE TEXT    NOT NULL
);


-- Table: PAYEE_V1
CREATE TABLE PAYEE_V1 (
    PAYEEID   INTEGER PRIMARY KEY,
    PAYEENAME TEXT    NOT NULL,
    CATEGID   NUMERIC,
    NUMBER    TEXT,
    WEBSITE   TEXT,
    NOTES     TEXT,
    ACTIVE    INTEGER,
    PATTERN   TEXT    DEFAULT ''
);


-- Table: REPORT_V1
CREATE TABLE REPORT_V1 (
    REPORTID        INTEGER NOT NULL
                            PRIMARY KEY,
    REPORTNAME      TEXT    COLLATE NOCASE
                            NOT NULL
                            UNIQUE,
    GROUPNAME       TEXT    COLLATE NOCASE,
    SQLCONTENT      TEXT,
    LUACONTENT      TEXT,
    TEMPLATECONTENT TEXT,
    DESCRIPTION     TEXT,
    ACTIVE          INTEGER
);


-- Table: SHAREINFO_V1
CREATE TABLE SHAREINFO_V1 (
    SHAREINFOID       INTEGER NOT NULL
                              PRIMARY KEY,
    CHECKINGACCOUNTID INTEGER NOT NULL,
    SHARENUMBER       NUMERIC,
    SHAREPRICE        NUMERIC,
    SHARECOMMISSION   NUMERIC,
    SHARELOT          TEXT
);


-- Table: SPLITTRANSACTIONS_V1
CREATE TABLE SPLITTRANSACTIONS_V1 (
    SPLITTRANSID     INTEGER PRIMARY KEY,
    TRANSID          NUMERIC NOT NULL,
    CATEGID          NUMERIC,
    SPLITTRANSAMOUNT NUMERIC,
    NOTES            TEXT
);


-- Table: STOCK_V1
CREATE TABLE STOCK_V1 (
    STOCKID       INTEGER PRIMARY KEY,
    HELDAT        NUMERIC,
    PURCHASEDATE  TEXT    NOT NULL,
    STOCKNAME     TEXT,
    SYMBOL        TEXT,
    NUMSHARES     NUMERIC,
    PURCHASEPRICE NUMERIC NOT NULL,
    NOTES         TEXT,
    CURRENTPRICE  NUMERIC NOT NULL,
    VALUE         NUMERIC,
    COMMISSION    NUMERIC
);


-- Table: STOCKHISTORY_V1
CREATE TABLE STOCKHISTORY_V1 (
    HISTID  INTEGER PRIMARY KEY,
    SYMBOL  TEXT    NOT NULL,
    DATE    TEXT    NOT NULL,
    VALUE   NUMERIC NOT NULL,
    UPDTYPE INTEGER,
    UNIQUE (
        SYMBOL,
        DATE
    )
);


-- Table: TRANSLINK_V1
CREATE TABLE TRANSLINK_V1 (
    TRANSLINKID       INTEGER NOT NULL
                              PRIMARY KEY,
    CHECKINGACCOUNTID INTEGER NOT NULL,
    LINKTYPE          TEXT    NOT NULL/* Asset, Stock */,
    LINKRECORDID      INTEGER NOT NULL
);


-- Index: IDX_ACCOUNTLIST_ACCOUNTTYPE
CREATE INDEX IDX_ACCOUNTLIST_ACCOUNTTYPE ON ACCOUNTLIST_V1 (
    ACCOUNTTYPE
);


-- Index: IDX_ASSETS_ASSETTYPE
CREATE INDEX IDX_ASSETS_ASSETTYPE ON ASSETS_V1 (
    ASSETTYPE
);


-- Index: IDX_ATTACHMENT_REF
CREATE INDEX IDX_ATTACHMENT_REF ON ATTACHMENT_V1 (
    REFTYPE,
    REFID
);


-- Index: IDX_BILLSDEPOSITS_ACCOUNT
CREATE INDEX IDX_BILLSDEPOSITS_ACCOUNT ON BILLSDEPOSITS_V1 (
    ACCOUNTID,
    TOACCOUNTID
);


-- Index: IDX_BUDGETSPLITTRANSACTIONS_TRANSID
CREATE INDEX IDX_BUDGETSPLITTRANSACTIONS_TRANSID ON BUDGETSPLITTRANSACTIONS_V1 (
    TRANSID
);


-- Index: IDX_BUDGETTABLE_BUDGETYEARID
CREATE INDEX IDX_BUDGETTABLE_BUDGETYEARID ON BUDGETTABLE_V1 (
    BUDGETYEARID
);


-- Index: IDX_BUDGETYEAR_BUDGETYEARNAME
CREATE INDEX IDX_BUDGETYEAR_BUDGETYEARNAME ON BUDGETYEAR_V1 (
    BUDGETYEARNAME
);


-- Index: IDX_CATEGORY_CATEGNAME
CREATE INDEX IDX_CATEGORY_CATEGNAME ON CATEGORY_V1 (
    CATEGNAME
);


-- Index: IDX_CATEGORY_CATEGNAME_PARENTID
CREATE INDEX IDX_CATEGORY_CATEGNAME_PARENTID ON CATEGORY_V1 (
    CATEGNAME,
    PARENTID
);


-- Index: IDX_CHECKINGACCOUNT
CREATE INDEX IDX_CHECKINGACCOUNT ON TRANSLINK_V1 (
    CHECKINGACCOUNTID
);


-- Index: IDX_CHECKINGACCOUNT_ACCOUNT
CREATE INDEX IDX_CHECKINGACCOUNT_ACCOUNT ON CHECKINGACCOUNT_V1 (
    ACCOUNTID,
    TOACCOUNTID
);


-- Index: IDX_CHECKINGACCOUNT_TRANSDATE
CREATE INDEX IDX_CHECKINGACCOUNT_TRANSDATE ON CHECKINGACCOUNT_V1 (
    TRANSDATE
);


-- Index: IDX_CURRENCYFORMATS_SYMBOL
CREATE INDEX IDX_CURRENCYFORMATS_SYMBOL ON CURRENCYFORMATS_V1 (
    CURRENCY_SYMBOL
);


-- Index: IDX_CURRENCYHISTORY_CURRENCYID_CURRDATE
CREATE INDEX IDX_CURRENCYHISTORY_CURRENCYID_CURRDATE ON CURRENCYHISTORY_V1 (
    CURRENCYID,
    CURRDATE
);


-- Index: IDX_CUSTOMFIELD_REF
CREATE INDEX IDX_CUSTOMFIELD_REF ON CUSTOMFIELD_V1 (
    REFTYPE
);


-- Index: IDX_CUSTOMFIELDDATA_REF
CREATE INDEX IDX_CUSTOMFIELDDATA_REF ON CUSTOMFIELDDATA_V1 (
    FIELDID,
    REFID
);


-- Index: IDX_INFOTABLE_INFONAME
CREATE INDEX IDX_INFOTABLE_INFONAME ON INFOTABLE_V1 (
    INFONAME
);


-- Index: IDX_LINKRECORD
CREATE INDEX IDX_LINKRECORD ON TRANSLINK_V1 (
    LINKTYPE,
    LINKRECORDID
);


-- Index: IDX_PAYEE_INFONAME
CREATE INDEX IDX_PAYEE_INFONAME ON PAYEE_V1 (
    PAYEENAME
);


-- Index: IDX_SHAREINFO
CREATE INDEX IDX_SHAREINFO ON SHAREINFO_V1 (
    CHECKINGACCOUNTID
);


-- Index: IDX_SPLITTRANSACTIONS_TRANSID
CREATE INDEX IDX_SPLITTRANSACTIONS_TRANSID ON SPLITTRANSACTIONS_V1 (
    TRANSID
);


-- Index: IDX_STOCK_HELDAT
CREATE INDEX IDX_STOCK_HELDAT ON STOCK_V1 (
    HELDAT
);


-- Index: IDX_STOCKHISTORY_SYMBOL
CREATE INDEX IDX_STOCKHISTORY_SYMBOL ON STOCKHISTORY_V1 (
    SYMBOL
);


-- Index: INDEX_REPORT_NAME
CREATE INDEX INDEX_REPORT_NAME ON REPORT_V1 (
    REPORTNAME
);


COMMIT TRANSACTION;
PRAGMA foreign_keys = on;

**Reply by vomikan** [reply]
Performance is negatively affected by a large number of payees.

**Reply by easyservices** [reply]
I have 414 payees in the table. What do you suggest?

**Reply by tactilis** [reply]
> I have 414 payees in the table.

That's a small number of payees.

Repeating my earlier test...
> I've just created a test database containing 30,000 transactions with a range of withdrawal and deposit values.
Filtering on an Amount Range of 15.00 to 100.00 generates 9,800 records and takes approx 0.5 seconds on my laptop from the moment of clicking OK to the the transaction list redraw being completed (measured on a screen video).

...but this time also creating 30000 distinct payees (a separate one for each transaction), the query time has increased from 0.5 seconds to 2 seconds.
